### ISSUE TITLE
Enhancement: [explicit-function-return-type] add option to allow direct function calls and allowed function name regex

### ISSUE THREAD

**Issue by marcindz88** [typescript-eslint/typescript-eslint#6295]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/explicit-function-return-type/

### Description

1. Function name macher can be used only to disable the rule on particular function names, In my opinion it should be possible to allow all functions that ends with particular string etc. This can be obtained by changing the function uncludes to regex match. However if that induces significant perfomance issue, it may be extracted to another option e.g. 'allowedRegexNames'

2. It is possible to disable rule for some kind of short function or those returning other functions directly, I found a usecase where it would be nice if i could disable the rule for functions which directly return other function calls e.g.NgRx selector factory, factories in general, or functions which use other functions but set some parameters, and want to use inferred type from inner function.

### Fail

```typescript
// options: [{ allowedNames: ['test.*'] } ],

export const foo = {
  tes() {
    return 0;
  },
};

// options: [{ allowFunctionContainingOnlyOtherFunctionCall: true } ],

const innerFunction = (a: number, b: number): number => 0;
const test = (a: number, b: number) => {
  const x = 10;
  return innerFunction(a);
};
```

### Pass

```typescript
// options: [{ allowedNames: ['test.*'] } ],

const test1 = () => {
  return;
};
export const foo = {
  test2() {
    return 0;
  },
};

// options: [{ allowFunctionContainingOnlyOtherFunctionCall: true } ],

function innerFunction1(a: number): number {
  return 0;
}
function test1(a: number, b: number) {
  return innerFunction1(a);
}

function innerFunction2(a: number): number {
  return 0;
}
const test2 = (a: number) => innerFunction2(a);

const innerFunction3 = (a: number): number => 0;
const test3 = (a: number) => innerFunction3(a);

const innerFunction4 = (a: number): number => 0;
const test4 = (a: number) => {
  return innerFunction4(a);
};
```

**Comment by bradzacher** [comment]
> It is possible to disable rule for some kind of short function or those returning other functions directly

Strongly disagree with this. The point of the rule is to define strict contracts. Other functions can have weak or no contract at all, or they may have complex contracts based on generic types.

Returning a function call isn't some magic thing that's better than just any old expression.

We won't be adding an option to allow this.

---

I don't think a broad regex option is a good idea. It makes it very, very easy to cause the rule to false negative and make your codebase inconsistent. Which is really bad.

For example your specified regex `test.*` would match a function name like `contestResult`, because you didn't prefix the regex with `^`.

It would be hard to specifically craft a regex that matches just the functions you care about.

For these reasons I don't think this is a good option to include in the rule.
